automata/fa/dfa.py

```diff
@@ -710,17 +710,159 @@ def maximum_word_length(self):
             return float('inf')
 
     @classmethod
-    def from_finite_language(cls, language, input_symbols):
+    def contains_substring(cls, input_symbols, substring):
+        """
+        Directly computes the minimal DFA recognizing strings containing the
+        given substring.
+        """
+        prefixes = [substring[:i] for i in range(len(substring))]
+
+        transitions = {prefix: dict() for prefix in prefixes}
+        transitions[substring] = {
+            symbol: substring for symbol in input_symbols
+        }
+
+        for prefix in prefixes:
+            prefix_dict = transitions.setdefault(prefix, dict())
+
+            for symbol in input_symbols:
+                # Look for next state after reading in the given input symbol
+                possible_suffix = prefix + symbol
+
+                # This while loop will always terminate, since the empty string is in the dict
+                while possible_suffix not in transitions:
+                    possible_suffix = possible_suffix[1:]
+
+                prefix_dict[symbol] = possible_suffix
+
+        return cls(
+            states=set(transitions.keys()),
+            input_symbols=input_symbols,
+            transitions=transitions,
+            initial_state='',
+            final_states={substring},
+        )
+
+    @classmethod
+    def contains_subsequence(cls, input_symbols, subsequence):
```

Comment:
@Tagl Per my comment above, can we rename this method to `from_subsequence`?

Reply:
Updated, see further discussion on other comment